## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#055fd16e-d8a7-4d94-8ccf-6b642bca0e5b) |
| mint-large-bucket.sh | :x: [more...](#f001d226-d853-41f4-af13-b4c1d8333268) |
| mint-dist-xl.sh | :x: [more...](#1fa0ae1a-432d-47e2-9b8e-c2861d06beba) |
| mint-gateway-azure.sh | :x: [more...](#2b02efc7-64a3-4709-87ec-41f678f589fb) |
| mint-zoned.sh | :x: [more...](#e7ef8c9b-5fdf-4ccd-ac61-00aa77515358) |
<a name="e7ef8c9b-5fdf-4ccd-ac61-00aa77515358"></a>
### 10025-0cca4c5/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30415
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 18e6636b888f:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 33 seconds
(8/15) Running minio-dotnet tests ... done in 55 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 49 seconds
{
  "args": {
    "bucketName": "minio-go-test-yfyojl0txoustsiv",
    "objectName": "a4ujcxy3a9s5d3q6ut3tvmu6x6ilkq",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 12383,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 33 seconds
(10/15) Running minio-js tests ... done in 50 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 17 seconds
(12/15) Running s3cmd tests ... done in 29 seconds
(13/15) Running s3select tests ... done in 10 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="2b02efc7-64a3-4709-87ec-41f678f589fb"></a>
### 10025-0cca4c5/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:32657
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5d12a16cab11:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 3 minutes and 26 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5f0a4c8b87882"
  },
  "duration": "40643",
  "status": "FAIL",
  "error": "Error executing \"CreateBucket\" on \"http://minio-c2.minio.io:32657/aws-sdk-php-5f0a4c8b87882\"; AWS HTTP error: Client error: `PUT http://minio-c2.minio.io:32657/aws-sdk-php-5f0a4c8b87882` resulted in a `409 Conflict` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>ContainerBeingDeleted</Code><Message>-&gt; github.com/Azure/azure-st (truncated...)\n ContainerBeingDeleted (client): -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.8.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=ContainerBeingDeleted) =====\nDescription=The specified container is being deleted. Try operation later.\nRequestId:f126db0e-001e-001e-29dc-573703000000\nTime:2020-07-11T23:38:01.1285647Z, Details: \n   Code: ContainerBeingDeleted\n   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5f0a4c8b87882?restype=container&timeout=1501\n   Authorization: REDACTED\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2020-07-11T23:28:33Z]\n   X-Ms-Client-Request-Id: [e20c4a7a-e300-464a-48c9-d7ab7292d841]\n   X-Ms-Date: [Sat, 11 Jul 2020 23:38:01 GMT]\n   X-Ms-Version: [2018-11-09]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The specified container is being deleted. Try operation later.\n   Content-Length: [252]\n   Content-Type: [application/xml]\n   Date: [Sat, 11 Jul 2020 23:38:00 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Error-Code: [ContainerBeingDeleted]\n   X-Ms-Request-Id: [f126db0e-001e-001e-29dc-573703000000]\n   X-Ms-Version: [2018-11-09]\n\n\n - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>ContainerBeingDeleted</Code><Message>-&gt; github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.8.0/azblob/zc_storage_error.go:42&#xA;===== RESPONSE ERROR (ServiceCode=ContainerBeingDeleted) =====&#xA;Description=The specified container is being deleted. Try operation later.&#xA;RequestId:f126db0e-001e-001e-29dc-573703000000&#xA;Time:2020-07-11T23:38:01.1285647Z, Details: &#xA;   Code: ContainerBeingDeleted&#xA;   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5f0a4c8b87882?restype=container&amp;timeout=1501&#xA;   Authorization: REDACTED&#xA;   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2020-07-11T23:28:33Z]&#xA;   X-Ms-Client-Request-Id: [e20c4a7a-e300-464a-48c9-d7ab7292d841]&#xA;   X-Ms-Date: [Sat, 11 Jul 2020 23:38:01 GMT]&#xA;   X-Ms-Version: [2018-11-09]&#xA;   --------------------------------------------------------------------------------&#xA;   RESPONSE Status: 409 The specified container is being deleted. Try operation later.&#xA;   Content-Length: [252]&#xA;   Content-Type: [application/xml]&#xA;   Date: [Sat, 11 Jul 2020 23:38:00 GMT]&#xA;   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]&#xA;   X-Ms-Error-Code: [ContainerBeingDeleted]&#xA;   X-Ms-Request-Id: [f126db0e-001e-001e-29dc-573703000000]&#xA;   X-Ms-Version: [2018-11-09]&#xA;&#xA;&#xA;</Message><BucketName>aws-sdk-php-5f0a4c8b87882</BucketName><Resource>/aws-sdk-php-5f0a4c8b87882</Resource><RequestId>1620D7042F958623</RequestId><HostId>f0ff50c4-067a-4349-bfc2-e28c84c6d569</HostId></Error>"
}
(3/15) Running aws-sdk-ruby tests ... done in 16 seconds
(4/15) Running awscli tests ... done in 2 minutes and 43 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 2 minutes and 56 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 31 seconds
(8/15) Running minio-go tests ... done in 4 minutes and 37 seconds
(9/15) Running minio-java tests ... done in 6 minutes and 5 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 32 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 44 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 21 seconds
(13/15) Running s3select tests ... done in 1 minutes and 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="055fd16e-d8a7-4d94-8ccf-6b642bca0e5b"></a>
### 10025-0cca4c5/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30285
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a625efe161eb:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 8 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 43 seconds
(8/15) Running minio-dotnet tests ... done in 31 seconds
(9/15) Running minio-go tests ... FAILED in 57 seconds
{
  "args": {
    "bucketName": "minio-go-test-k2stzttgqfuhhlof",
    "objectName": "e00ubjsuvsq15lo2na4q9fhjokfzi2",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 5124,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 43 seconds
(10/15) Running minio-js tests ... done in 46 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 28 seconds
(12/15) Running s3cmd tests ... done in 17 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="1fa0ae1a-432d-47e2-9b8e-c2861d06beba"></a>
### 10025-0cca4c5/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31379
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c3d1d10ef131:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 10 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 46 seconds
(8/15) Running minio-dotnet tests ... done in 43 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 34 seconds
{
  "args": {
    "bucketName": "minio-go-test-zshctk1uzh962djq",
    "objectName": "20x3zsnqm3q5okjutppnc0anrtflx4",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 3528,
  "error": "The bucket you tried to delete is not empty. You must delete all versions in the bucket.",
  "function": "RemoveObjectsWithOptions(bucketName, objectsCh, opts)",
  "message": "Cleanup failed",
  "name": "minio-go: testRemoveObjectsWithOptions",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 1 minutes and 11 seconds
(10/15) Running minio-js tests ... done in 56 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 24 seconds
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="f001d226-d853-41f4-af13-b4c1d8333268"></a>
### 10025-0cca4c5/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30892
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a09e2cb294a5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-l59zrmvmhcbcbp",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 366085,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 1620D7276782DB91, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-l59zrmvmhcbcbp",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 366085,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 1620D7276782DB91, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-php tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-l59zrmvmhcbcbp",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 366085,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 1620D7276782DB91, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-ruby tests ... FAILED in 3 seconds
{
  "name": "aws-sdk-ruby",
  "function": "listBuckets",
  "args": {},
  "duration": 2128.41,
  "error": "Server not initialized, please try again.",
  "status": "FAIL"
}
(1/15) Running awscli tests ... FAILED in 13 seconds
{
  "name": "awscli",
  "duration": 11548,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}
(1/15) Running healthcheck tests ... done in 1 seconds
(2/15) Running mc tests ... FAILED in 0 seconds
(2/15) Running minio-dotnet tests ... done in 6 seconds
(3/15) Running minio-go tests ... FAILED in 1 minutes and 26 seconds
{
  "args": {
    "bucketName": "minio-go-test-30bvsc6ovwhnerdd",
    "region": "eu-west-1"
  },
  "duration": 86519,
  "error": "Server not initialized, please try again.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(3/15) Running minio-java tests ... FAILED in 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(MakeBucketArgs args)",
  "duration": 230,
  "status": "FAIL",
  "error": "java.lang.IllegalArgumentException: unknown error code string 'XMinioServerNotInitialized' >>> [io.minio.Xml.unmarshal(Xml.java:68), io.minio.MinioClient.execute(MinioClient.java:1050), io.minio.MinioClient.makeBucket(MinioClient.java:3773), FunctionalTest.makeBucket_test1(FunctionalTest.java:435), FunctionalTest.runTests(FunctionalTest.java:3772), FunctionalTest.main(FunctionalTest.java:4056)]"
}
(3/15) Running minio-js tests ... FAILED in 2 seconds
{
  "name": "minio-js",
  "function": "makeBucket(bucketName, region, cb)",
  "args": "bucketName:minio-js-test-813e360e-a868-4779-8852-0bc2010b2a29-region, region:us-east-1",
  "duration": 41,
  "status": "FAIL",
  "error": "S3Error: Server not initialized, please try again. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(3/15) Running minio-py tests ... FAILED in 10 minutes and 28 seconds
{
  "name": "minio-py:test_list_objects",
  "function": "list_objects(bucket_name, prefix, recursive)",
  "args": {
    "bucket_name": "minio-py-test-4b716b0f-8112-44b7-b10b-2add0b7ab1c7",
    "object_name": "990bcf09-5fe2-4ee8-98d2-83b8c98c4251",
    "recursive": true
  },
  "duration": 601927,
  "message": "HTTPConnectionPool(host='minio-dev6.minio.io', port=30892): Max retries exceeded with url: /minio-py-test-4b716b0f-8112-44b7-b10b-2add0b7ab1c7/?prefix= (Caused by ResponseError('too many 503 error responses',))",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1035, in test_list_objects\n    for obj in objects:\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1036, in list_objects\n    headers=headers)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2000, in _url_open\n    preload_content=preload_content)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/poolmanager.py\", line 336, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 828, in urlopen\n    **response_kw\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py\", line 805, in urlopen\n    retries = retries.increment(method, url, response=response, _pool=self)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/util/retry.py\", line 439, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minio-dev6.minio.io', port=30892): Max retries exceeded with url: /minio-py-test-4b716b0f-8112-44b7-b10b-2add0b7ab1c7/?prefix= (Caused by ResponseError('too many 503 error responses',))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2085, in main\n    test_list_objects(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1042, in test_list_objects\n    raise Exception(err)\nException: HTTPConnectionPool(host='minio-dev6.minio.io', port=30892): Max retries exceeded with url: /minio-py-test-4b716b0f-8112-44b7-b10b-2add0b7ab1c7/?prefix= (Caused by ResponseError('too many 503 error responses',))\n",
  "status": "FAIL"
}
(3/15) Running s3cmd tests ... FAILED in 45 seconds
{
  "name": "s3cmd",
  "duration": "45524",
  "function": "test_make_bucket",
  "status": "FAIL",
  "error": "WARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 3 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 6 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 9 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 12 sec...\nWARNING: Retrying failed request: / (503 (XMinioServerNotInitialized): Server not initialized, please try again.)\nWARNING: Waiting 15 sec...\nERROR: S3 error: 503 (XMinioServerNotInitialized): Server not initialized, please try again."
}
(3/15) Running s3select tests ... done in 1 hours 10 minutes and 14 seconds
(4/15) Running security tests ... done in 0 seconds

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/api-router.go
+++ b/cmd/api-router.go
@@ -291,18 +291,14 @@ func registerAPIRouter(router *mux.Router, encryptionEnabled, allowSSEKMS bool) 
 	apiRouter.Methods(http.MethodGet).Path(SlashSeparator + SlashSeparator).HandlerFunc(
 		maxClients(collectAPIStats("listbuckets", httpTraceAll(api.ListBucketsHandler))))
 
-	// Supports cors only for S3 handlers
-	apiRouter.Methods(http.MethodOptions).HandlerFunc(
-		maxClients(collectAPIStats("cors", httpTraceAll(corsHandlerFunc()))))
-
 	// If none of the routes match add default error handler routes
 	apiRouter.NotFoundHandler = http.HandlerFunc(collectAPIStats("notfound", httpTraceAll(errorResponseHandler)))
 	apiRouter.MethodNotAllowedHandler = http.HandlerFunc(collectAPIStats("methodnotallowed", httpTraceAll(errorResponseHandler)))
 
 }
 
-// setCorsHandler handler for CORS (Cross Origin Resource Sharing)
-func corsHandlerFunc() http.HandlerFunc {
+// corsHandler handler for CORS (Cross Origin Resource Sharing)
+func corsHandler(handler http.Handler) http.Handler {
 	commonS3Headers := []string{
 		xhttp.Date,
 		xhttp.ETag,
@@ -318,7 +314,7 @@ func corsHandlerFunc() http.HandlerFunc {
 		"*",
 	}
 
-	c := cors.New(cors.Options{
+	return cors.New(cors.Options{
 		AllowOriginFunc: func(origin string) bool {
 			for _, allowedOrigin := range globalAPIConfig.getCorsAllowOrigins() {
 				if wildcard.MatchSimple(allowedOrigin, origin) {
@@ -339,7 +335,5 @@ func corsHandlerFunc() http.HandlerFunc {
 		AllowedHeaders:   commonS3Headers,
 		ExposedHeaders:   commonS3Headers,
 		AllowCredentials: true,
-	})
-
-	return c.HandlerFunc
+	}).Handler(handler)
 }

--- a/cmd/gateway-main.go
+++ b/cmd/gateway-main.go
@@ -267,7 +267,7 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 	}
 
 	httpServer := xhttp.NewServer([]string{globalCLIContext.Addr},
-		criticalErrorHandler{router}, getCert)
+		criticalErrorHandler{corsHandler(router)}, getCert)
 	httpServer.BaseContext = func(listener net.Listener) context.Context {
 		return GlobalContext
 	}

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -466,7 +466,7 @@ func serverMain(ctx *cli.Context) {
 		}
 	}()
 
-	httpServer := xhttp.NewServer([]string{globalMinioAddr}, criticalErrorHandler{handler}, getCert)
+	httpServer := xhttp.NewServer([]string{globalMinioAddr}, criticalErrorHandler{corsHandler(handler)}, getCert)
 	httpServer.ErrorLog = log.New(pw, "", 0)
 	httpServer.BaseContext = func(listener net.Listener) context.Context {
 		return GlobalContext

